METAL-1038 Remove baremetal dependency on MachineAPI capability

eranco74 opened the pull request:
There are use cases for using the baremetal platform and having the baremetal capability while disabling MAPI (for example to use CAPI ).
Currently, the clusterversion CRD validates that baremetal capability isn't enabled while MachineAPI capability is disabled, the validation exists because CBO will crash if the Machine CRD doesn't exist on the cluster.
CBO is usable on many platforms, and depends on MAPI on all of them, when it only really needs it for baremetal IPI.
This [PR](https://github.com/openshift/cluster-baremetal-operator/pull/423) in CBO should allow it to run properly when MachineAPI is disalbed

eranco74 commented:
/cc @JoelSpeed 
/cc @zaneb
/cc @dtantsur

zaneb commented on config/v1/types_cluster_version.go:
```diff
@@ -647,8 +646,8 @@ type ComponentOverride struct {
-// +kubebuilder:validation:XValidation:rule="has(self.architecture) && has(self.image) ? (self.architecture == '' || self.image == '') : true",message="cannot set both Architecture and Image"
-// +kubebuilder:validation:XValidation:rule="has(self.architecture) && self.architecture != '' ? self.version != '' : true",message="Version must be set if Architecture is set"
+// +kubebuilder:validation:XValidation:rule="has(self.architecture) && has(self.image) ? (self.architecture == ” || self.image == ”) : true",message="cannot set both Architecture and Image"
+// +kubebuilder:validation:XValidation:rule="has(self.architecture) && self.architecture != ” ? self.version != ” : true",message="Version must be set if Architecture is set"
```
This doesn't look valid

JoelSpeed replied:
Yeah I think this is your IDE trying to be smart, but in fact, breaking the quoting, these lines will need to be reverted

eranco74 replied:
:confused: 
Thanks, unsure how that happend

eranco74 commented:
@lshilin-r is the QE for this epic.
These are the PRs required to test this change:

* [openshift installer](https://github.com/openshift/installer/pull/8521):
Allow disabling MAPI while installing baremetall platform (with baremetal capability enabled)
Update the openshift installer to not render BMH CR when MAPI is disabled.
 
* [api](https://github.com/openshift/api/pull/1916):
Allow creating clusterversion CR with `baremetal` capability without the `MachineAPI` capability 

* [cluster-version-operator](https://github.com/openshift/cluster-version-operator/pull/1051):
Use the updated API

* [cluster-baremetal-operator](https://github.com/openshift/cluster-baremetal-operator/pull/423):
Update the provisioning controller not to crash in case the Machine CRD isn't registered

eranco74 commented:
note that I've tested the configuration and successfully installed OCP with baremetall platform without MachineAPI using assisted installer.
Steps:

1. Build release image with the above mentioned PRs, using cluster bot
2. Setup assisted test env using [assisted-test-infra](https://github.com/openshift/assisted-test-infra).
3. Updated assisted-service configmap to use this release image for the installaltion
4. Restarted the assisted service deployment to reload the configuration
5.  Created a cluster in assisted:
     `make deploy_nodes OPENSHIFT_VERSION=4.16  NUM_MASTERS=3 NUM_WORKERS=0`
6. Updated the cluster capabilities (using install config override):
    ```
    export ASSISTED_SERVICE_IP=<replace this>
export ASSISTED_SERVICE_PORT=6000
export CLUSTER_UUID=<replace this>
```sh
curl --request PATCH --header "Content-Type: application/json"     --data '"{\"capabilities\": {\"baselineCapabilitySet\": \"None\", \"additionalEnabledCapabilities\": [\"openshift-samples\", \"marketplace\", \"Console\", \"baremetal\", \"Insights\", \"Storage\", \"NodeTuning\", \"CSISnapshot\", \"OperatorLifecycleManager\", \"Ingress\"]}}"' "http://$ASSISTED_SERVICE_IP:$ASSISTED_SERVICE_PORT/api/assisted-install/v2/clusters/$CLUSTER_UUID/install-config"
```
6. Triggered the insallation
7. Verified the cluster installation completed successfully and that MachineAPI isn't running:
```
[root@rdu-infra-edge-05 assisted-test-infra]# kubectl --kubeconfig ./kubeconfig  get clusterversion
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         3m17s   Cluster version is 4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest
[root@rdu-infra-edge-05 assisted-test-infra]# kubectl --kubeconfig ./kubeconfig  get co
NAME                                       VERSION                                                   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
authentication                             4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      3m49s   
baremetal                                  4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      29m     
config-operator                            4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      30m     
console                                    4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      10m     
csi-snapshot-controller                    4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      30m     
dns                                        4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      29m     
etcd                                       4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      29m     
ingress                                    4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      20m     
insights                                   4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      24m     
kube-apiserver                             4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      27m     
kube-controller-manager                    4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      27m     
kube-scheduler                             4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      27m     
kube-storage-version-migrator              4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      30m     
machine-approver                           4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      30m     
machine-config                             4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      29m     
marketplace                                4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      29m     
monitoring                                 4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      11m     
network                                    4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      31m     
node-tuning                                4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      17m     
openshift-apiserver                        4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      14m     
openshift-controller-manager               4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      26m     
openshift-samples                          4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      21m     
operator-lifecycle-manager                 4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      29m     
operator-lifecycle-manager-catalog         4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      29m     
operator-lifecycle-manager-packageserver   4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      25m     
service-ca                                 4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      30m     
storage                                    4.16.0-0.ci.test-2024-06-18-154755-ci-ln-mkd6npb-latest   True        False         False      30m  
```

eranco74 commented:
Also verified there's no regression and that IPI baremetal installation works:
```
oc get node,machine,bmh -A
NAME            STATUS   ROLES                  AGE   VERSION
node/master-0   Ready    control-plane,master   49m   v1.30.1+9798e19
node/master-1   Ready    control-plane,master   46m   v1.30.1+9798e19
node/master-2   Ready    control-plane,master   45m   v1.30.1+9798e19
node/worker-0   Ready    worker                 17m   v1.30.1+9798e19
node/worker-1   Ready    worker                 17m   v1.30.1+9798e19
node/worker-2   Ready    worker                 17m   v1.30.1+9798e19

NAMESPACE               NAME                                                       PHASE     TYPE   REGION   ZONE   AGE
openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-master-0         Running                          58m
openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-master-1         Running                          58m
openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-master-2         Running                          58m
openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-worker-0-2rl4x   Running                          43m
openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-worker-0-8n2d2   Running                          43m
openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-worker-0-qpgbf   Running                          43m

NAMESPACE               NAME                                      STATE         CONSUMER                      ONLINE   ERROR   AGE
openshift-machine-api   baremetalhost.metal3.io/ostest-master-0   provisioned   ostest-h5mtv-master-0         true             58m
openshift-machine-api   baremetalhost.metal3.io/ostest-master-1   provisioned   ostest-h5mtv-master-1         true             58m
openshift-machine-api   baremetalhost.metal3.io/ostest-master-2   provisioned   ostest-h5mtv-master-2         true             58m
openshift-machine-api   baremetalhost.metal3.io/ostest-worker-0   provisioned   ostest-h5mtv-worker-0-8n2d2   true             58m
openshift-machine-api   baremetalhost.metal3.io/ostest-worker-1   provisioned   ostest-h5mtv-worker-0-qpgbf   true             58m
openshift-machine-api   baremetalhost.metal3.io/ostest-worker-2   provisioned   ostest-h5mtv-worker-0-2rl4x   true             58m
```

JoelSpeed commented:
/lgtm

Based on above testing, I'm happy for this to move forward

lshilin-r commented:
adding Elai

On Tue, Jun 25, 2024 at 10:40 AM Eran Cohen ***@***.***>
wrote:

> Also verified there's no regression and that IPI baremetal installation
> works:
>
> oc get node,machine,bmh -A
> NAME            STATUS   ROLES                  AGE   VERSION
> node/master-0   Ready    control-plane,master   49m   v1.30.1+9798e19
> node/master-1   Ready    control-plane,master   46m   v1.30.1+9798e19
> node/master-2   Ready    control-plane,master   45m   v1.30.1+9798e19
> node/worker-0   Ready    worker                 17m   v1.30.1+9798e19
> node/worker-1   Ready    worker                 17m   v1.30.1+9798e19
> node/worker-2   Ready    worker                 17m   v1.30.1+9798e19
>
> NAMESPACE               NAME                                                       PHASE     TYPE   REGION   ZONE   AGE
> openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-master-0         Running                          58m
> openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-master-1         Running                          58m
> openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-master-2         Running                          58m
> openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-worker-0-2rl4x   Running                          43m
> openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-worker-0-8n2d2   Running                          43m
> openshift-machine-api   machine.machine.openshift.io/ostest-h5mtv-worker-0-qpgbf   Running                          43m
>
> NAMESPACE               NAME                                      STATE         CONSUMER                      ONLINE   ERROR   AGE
> openshift-machine-api   baremetalhost.metal3.io/ostest-master-0   provisioned   ostest-h5mtv-master-0         true             58m
> openshift-machine-api   baremetalhost.metal3.io/ostest-master-1   provisioned   ostest-h5mtv-master-1         true             58m
> openshift-machine-api   baremetalhost.metal3.io/ostest-master-2   provisioned   ostest-h5mtv-master-2         true             58m
> openshift-machine-api   baremetalhost.metal3.io/ostest-worker-0   provisioned   ostest-h5mtv-worker-0-8n2d2   true             58m
> openshift-machine-api   baremetalhost.metal3.io/ostest-worker-1   provisioned   ostest-h5mtv-worker-0-qpgbf   true             58m
> openshift-machine-api   baremetalhost.metal3.io/ostest-worker-2   provisioned   ostest-h5mtv-worker-0-2rl4x   true             58m
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/api/pull/1916#issuecomment-2188198714>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AO5F5SGMFD64XSQ3UI2PA2DZJENGJAVCNFSM6AAAAABI23MTTSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCOBYGE4TQNZRGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Best Regards,
Lubov
